### PR TITLE
Feat(generator): boost emissions

### PR DESCRIPTION
**Pull request by P-Yevhenii**

### CODE REVIEW

File: contracts/tokenomics/generator/src/contract.rs [hunk 1]
```diff
@@ -1567,6 +1625,9 @@ pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> Result<Binary, ContractErro
     match msg {
         QueryMsg::ActivePoolLength {} => Ok(to_binary(&active_pool_length(deps)?)?),
         QueryMsg::PoolLength {} => Ok(to_binary(&pool_length(deps)?)?),
+        QueryMsg::EmissionRewards { lp_token, user } => {
+            Ok(to_binary(&query_emission_rewards(deps, lp_token, user)?)?)
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/contract.rs:1629]: I think this value does not make sense. We need to know users' boost at the current period which is the value from the formula _min(0.4 + 0.6 * (w_i/W), 1)._ where w_i - user VP and W - total VP.

File: contracts/tokenomics/generator/src/state.rs [hunk 2]
```diff
@@ -16,6 +16,8 @@ pub struct UserInfo {
     pub reward_debt: Uint128,
     /// Proxy reward amount a user already received or is not eligible for; used for proper reward calculation
     pub reward_debt_proxy: Uint128,
+    /// The amount of user emissions
+    pub emission_amount: Uint128,
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/state.rs:20]: We do not need to store this value. Store users' boost instead.

File: contracts/tokenomics/generator/tests/integration.rs [hunk 3]
```diff
@@ -1,5 +1,6 @@
 use astroport::asset::{native_asset_info, token_asset_info, AssetInfo, PairInfo};
```
Comment (epanchee) [inline contracts/tokenomics/generator/tests/integration.rs:1]: Revert file mode back to 644.

File: contracts/tokenomics/generator/tests/integration.rs [hunk 4]
```diff
@@ -49,6 +56,243 @@ struct PoolWithProxy {
...
+        &helper_controller.generator,
+        &lp_eur_usd,
+        &user1,
+        0,
+    );
+
+    check_token_balance(&mut app, &lp_cny_eur, &helper_controller.generator, 10);
+    check_token_balance(&mut app, &lp_eur_usd, &helper_controller.generator, 10);
+
+    app.update_block(|bi| next_block(bi));
```
Comment (epanchee) [inline contracts/tokenomics/generator/tests/integration.rs:293]: I think we do not need this line.

File: contracts/tokenomics/generator/src/contract.rs (hunk 1 above)
Comment (P-Yevhenii) [reply contracts/tokenomics/generator/src/contract.rs:1629]: Hm. It returns last calculated emission amount. If we'll use formula^ for calculate emission in real time, it will be wrong, because if we do nothing we have static emission amount calculated before and it changes only after next actions. Let's talk about it tomorrow.

File: contracts/tokenomics/generator/src/contract.rs [hunk 5]
```diff
@@ -279,6 +281,42 @@ pub fn execute(
     }
 }
 
+/// ## Description
+/// Update user lp emission in specified generators
+fn check_points(
```
[the remark this answers is not in the document]
Comment (P-Yevhenii) [reply contracts/tokenomics/generator/src/contract.rs:286]: Sure. Done.

File: contracts/tokenomics/generator/src/contract.rs [hunk 6]
```diff
@@ -279,6 +281,42 @@ pub fn execute(
...
+    mut deps: DepsMut,
+    env: Env,
+    user: String,
+    generators: Vec<String>,
+) -> Result<Response, ContractError> {
+    let user_addr = addr_validate_to_lower(deps.api, &user)?;
+    let config = CONFIG.load(deps.storage)?;
+
+    if generators.len() > GENERATORS_LIMIT as usize {
```
[the remark this answers is not in the document]
Comment (P-Yevhenii) [reply contracts/tokenomics/generator/src/contract.rs:295]: It is specified in the task.

File: contracts/tokenomics/generator/src/utils.rs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        generator.clone(),
+        env.contract.address.clone(),
+    )?;
+
+    let user_share_emission = Decimal::from_str("0.4")?.checked_mul(user_info.amount)?;
+    let total_share_emission = Decimal::from_str("0.6")?.checked_mul(total_balance)?;
+    let vx_share_emission = Decimal::from_ratio(user_vp, total_vp);
+
+    let virtual_amount =
+        user_share_emission.checked_add(vx_share_emission.checked_mul(total_share_emission)?)?;
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/utils.rs:53]: We do not need checked functions here because virtual_amount is composed of 0.4*user_info.amount (which is small part of total_share_emission) and at most 0.6 * total_share_emission. The maximum possible value here is total_share_emission which fits in Uint128 if user_info.amount = total_share_emission which is highly unlikely possible.

File: contracts/tokenomics/generator/tests/integration.rs [hunk 8]
```diff
@@ -50,6 +57,272 @@ struct PoolWithProxy {
...
+    // check user2's ASTRO balance
+    check_token_balance(
+        &mut app,
+        &helper_controller.escrow_helper.astro_token,
+        &user2,
+        0,
+    );
+
+    app.next_block(WEEK);
+    app.update_block(|bi| bi.time = bi.time.plus_seconds(WEEK));
```
Comment (epanchee) [inline contracts/tokenomics/generator/tests/integration.rs:189]: app.next_block is just helper over app.update_block. If you want to shift by 2 weeks here simply use app.next_block(2*WEEK). If you do not need this then just remove this line (app.update_block(...)...).

File: contracts/tokenomics/generator/tests/integration.rs [hunk 9]
```diff
@@ -868,8 +1173,13 @@ fn generator_with_mirror_reward_proxy() {
...
     );
 
-    let generator_instance =
-        instantiate_generator(&mut app, &factory_instance, &astro_token_instance, None);
+    let generator_instance = instantiate_generator(
+        &mut app,
+        &factory_instance,
+        &astro_token_instance,
+        None,
+        None, //Some(OWNER.to_string()),
```
Comment (epanchee) [inline contracts/tokenomics/generator/tests/integration.rs:1181]: Do we need this comment?

File: packages/astroport/src/generator.rs [hunk 10]
```diff
@@ -151,6 +157,15 @@ pub enum ExecuteMsg {
     DeactivatePools {
         pair_types: Vec<PairType>,
     },
+    /// Updates the boost emissions for specified user and generators
+    CheckpointUserBoost {
+        generators: Vec<String>,
+    },
+    /// Kick a staker for specified generators if they abuse their boost
```
Comment (epanchee) [inline packages/astroport/src/generator.rs:164]: This is not about abusing but community-driven regulations

File: contracts/tokenomics/generator/src/contract.rs [hunk 11]
```diff
@@ -171,6 +188,10 @@ pub fn execute(
     msg: ExecuteMsg,
 ) -> Result<Response, ContractError> {
     match msg {
+        ExecuteMsg::KickUserBoost { user, generators } => kick_user_boost(deps, user, generators),
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/contract.rs:191]: Imo, 'Kick a user' just means recalculate his virtual amount and if it becomes 0 - do not pay him reward from now. This could be done by usual business logic in the Checkpoint function. So, I think we do not need this separate function.

File: contracts/tokenomics/generator/src/utils.rs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+/// Calculates emission boost amount for specified user and generator
+///
+/// **b_u = min(0.4 * b_u + 0.6 * S * (w_i / W), b_u)**
+///
+/// - b_u is the amount of LP tokens a user staked in a generator
+///
+/// - S is the total amount of LP tokens staked in a generator
+/// - w_i is a user’s current vxASTRO balance
+/// - W is the total amount of vxASTRO
+pub(crate) fn update_virtual_amount(
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/utils.rs:20]: I propose to include this logic in state.rs: update_user_balance() function.

**Review by epanchee — CHANGES_REQUESTED** [review]

File: contracts/tokenomics/generator/src/contract.rs [hunk 13]
```diff
@@ -1947,6 +2061,28 @@ pub fn query_deposit(deps: Deps, lp_token: String, user: String) -> Result<Uint1
...
+pub fn query_virtual_amount(
+    deps: Deps,
+    lp_token: String,
+    user: String,
+) -> Result<Uint128, ContractError> {
+    let lp_token = addr_validate_to_lower(deps.api, &lp_token)?;
+    let user = addr_validate_to_lower(deps.api, &user)?;
+
+    let user_info = USER_INFO
+        .load(deps.storage, (&lp_token, &user))
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/contract.rs:2081]: Don't use .load(). Use .compatible_load() only.

File: contracts/tokenomics/generator/src/contract.rs [hunk 14]
```diff
@@ -2012,29 +2148,25 @@ pub fn pending_token(
             }
         }
         None => {
-            lp_supply = query_token_balance(
-                &deps.querier,
-                lp_token.clone(),
-                env.contract.address.clone(),
-            )?;
+            lp_supply = pool.total_virtual_supply;
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/contract.rs:2151]: You need to use this value for pools with proxy as well

File: contracts/tokenomics/generator/src/contract.rs [hunk 15]
```diff
@@ -2012,29 +2148,25 @@ pub fn pending_token(
...
         let share = Decimal::from_ratio(token_rewards, lp_supply);
-        acc_per_share = pool.accumulated_rewards_per_share.checked_add(share)?;
+        acc_per_share = pool.reward_global_index.checked_add(share)?;
     }
 
-    let pending = acc_per_share
-        .checked_mul(user_info.amount)?
-        .checked_sub(user_info.reward_debt)?;
+    // we should calculate rewards by virtual amount
+    let pending_rewards =
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/contract.rs:2165]: Why did you rename it to pending_rewards?

File: contracts/tokenomics/generator/src/contract.rs [hunk 16]
```diff
@@ -2060,6 +2192,7 @@ fn query_config(deps: Deps) -> Result<ConfigResponse, ContractError> {
         generator_controller: config.generator_controller,
         active_pools: config.active_pools,
         blocked_tokens_list: config.blocked_tokens_list,
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/contract.rs:2194]: I think we need to add checkpoint_generator_limit here as well.

File: contracts/tokenomics/generator/src/contract.rs [hunk 17]
```diff
@@ -2433,11 +2567,12 @@ pub fn migrate(mut deps: DepsMut, env: Env, msg: MigrateMsg) -> Result<Response,
                         has_asset_rewards: false,
                         accumulated_proxy_rewards_per_share,
                         orphan_proxy_rewards,
-                        accumulated_rewards_per_share: pool_info_v100.accumulated_rewards_per_share,
                         last_reward_block: pool_info_v100.last_reward_block,
                         proxy_reward_balance_before_update: pool_info_v100
                             .proxy_reward_balance_before_update,
                         reward_proxy: pool_info_v100.reward_proxy,
+                        reward_global_index: pool_info_v100.accumulated_rewards_per_share,
+                        total_virtual_supply: Default::default(),
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/contract.rs:2575]: Total virtual supply should be set to 0.4 of current total supply, isn't it?

File: contracts/tokenomics/generator/src/contract.rs [hunk 18]
```diff
@@ -2521,16 +2657,66 @@ pub fn migrate(mut deps: DepsMut, env: Env, msg: MigrateMsg) -> Result<Response,
...
+                let keys = POOL_INFO
+                    .keys(deps.storage, None, None, cosmwasm_std::Order::Ascending {})
+                    .map(|v| String::from_utf8(v).map_err(StdError::from))
+                    .collect::<Result<Vec<String>, StdError>>()?;
+
+                for key in keys {
+                    let pool_info_v130 = migration::POOL_INFOV130
+                        .load(deps.storage, &Addr::unchecked(key.clone()))?;
+
+                    let mut accumulated_proxy_rewards_per_share = Default::default();
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/contract.rs:2683]: We must not migrate all this proxy related stuff since it should be already set in v.1.3.0.

File: contracts/tokenomics/generator/src/migration.rs [hunk 19]
```diff
@@ -35,7 +35,7 @@ pub struct PoolInfoV110 {
     pub alloc_point: Uint64,
     /// Accumulated amount of reward per share unit. Used for reward calculations
     pub last_reward_block: Uint64,
-    pub accumulated_rewards_per_share: Decimal,
+    pub reward_global_index: Decimal,
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/migration.rs:38]: Never change structures from old migrations. It may break migration.

File: contracts/tokenomics/generator/src/migration.rs [hunk 20]
```diff
@@ -54,7 +55,7 @@ pub const POOL_INFOV110: Map<&Addr, PoolInfoV110> = Map::new("pool_info");
 pub struct PoolInfoV120 {
     /// Accumulated amount of reward per share unit. Used for reward calculations
     pub last_reward_block: Uint64,
-    pub accumulated_rewards_per_share: Decimal,
+    pub reward_global_index: Decimal,
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/migration.rs:58]: Revert this structure too.

File: contracts/tokenomics/generator/src/migration.rs [hunk 21]
```diff
@@ -64,11 +65,32 @@ pub struct PoolInfoV120 {
...
 }
 
 /// Stores the contract config(V1.2.0) at the given key
 pub const POOL_INFOV120: Map<&Addr, PoolInfoV120> = Map::new("pool_info");
 
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct PoolInfoV130 {
+    /// Accumulated amount of reward per share unit. Used for reward calculations
+    pub last_reward_block: Uint64,
+    pub reward_global_index: Decimal,
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/migration.rs:78]: This structre does not have this field at version 1.3.0

File: contracts/tokenomics/generator/src/migration.rs [hunk 22]
```diff
@@ -64,11 +65,32 @@ pub struct PoolInfoV120 {
...
+    /// the reward proxy contract
+    pub reward_proxy: Option<Addr>,
+    pub accumulated_proxy_rewards_per_share: Decimal,
+    /// for calculation of new proxy rewards
+    pub proxy_reward_balance_before_update: Uint128,
+    /// the orphan proxy rewards which are left by emergency withdrawals
+    pub orphan_proxy_rewards: Uint128,
+    /// The pool has assets giving additional rewards
+    pub has_asset_rewards: bool,
+    pub total_virtual_supply: Uint128,
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/migration.rs:88]: Remove this field

File: contracts/tokenomics/generator/src/migration.rs [hunk 23]
```diff
@@ -64,11 +65,32 @@ pub struct PoolInfoV120 {
...
+}
+
+/// Stores the contract config(V1.3.0) at the given key
+pub const POOL_INFOV130: Map<&Addr, PoolInfoV130> = Map::new("pool_info");
+
 /// This structure describes the main control config of generator.
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
 pub struct ConfigV100 {
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/migration.rs:96]: Revert this structure

File: contracts/tokenomics/generator/src/migration.rs [hunk 24]
```diff
@@ -196,7 +219,83 @@ pub fn migrate_configs_to_v130(storage: &mut dyn Storage) -> StdResult<()> {
...
+        start_block: cfg_130.start_block,
+        allowed_reward_proxies: cfg_130.allowed_reward_proxies,
+        vesting_contract: cfg_130.vesting_contract,
+        active_pools: cfg_130.active_pools,
+        guardian: cfg_130.guardian,
+        generator_limit: None,
+        blocked_tokens_list: cfg_130.blocked_tokens_list,
+    };
+
+    if let Some(voting_escrow) = msg.voting_escrow {
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/migration.rs:290]: You can set these values during Config {} structure creation

File: contracts/tokenomics/generator/src/state.rs [hunk 25]
```diff
@@ -106,22 +115,33 @@ pub trait CompatibleLoader<K, R> {
...
+            let user_index = pool_info.reward_global_index
+                - Decimal::from_ratio(current_reward, old_user_info.amount);
+
             let user_info = UserInfoV2 {
-                amount: user_info.amount,
-                reward_debt: user_info.reward_debt,
+                amount: old_user_info.amount,
+                reward_user_index: user_index,
                 reward_debt_proxy,
+                virtual_amount: old_user_info.amount,
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/state.rs:144]: This should be set to 0.4 * old_user_info.amount

File: contracts/tokenomics/generator/src/state.rs [hunk 26]
```diff
@@ -218,3 +236,54 @@ pub fn update_proxy_asset(deps: DepsMut, proxy_addr: &Addr) -> StdResult<()> {
...
+    let user_virtual_share = user_info.amount.multiply_ratio(4u128, 10u128);
+    let total_virtual_share = pool.total_virtual_supply.multiply_ratio(6u128, 10u128);
+
+    let vx_share_emission = if !total_vp.is_zero() {
+        Decimal::from_ratio(user_vp, total_vp)
+    } else {
+        Decimal::zero()
+    };
+
+    Ok(user_virtual_share.checked_add(vx_share_emission.checked_mul(total_virtual_share)?)?)
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/state.rs:273]: We do not need checked functions here. I already wrote detailed explanations before. Checked functions are expensive because they are computed within *256 types. Do not use checked functions if calculations will never throw Overflow.

File: contracts/tokenomics/generator/src/state.rs [hunk 27]
```diff
@@ -218,3 +236,54 @@ pub fn update_proxy_asset(deps: DepsMut, proxy_addr: &Addr) -> StdResult<()> {
...
+    let mut user_vp = Uint128::zero();
+    let mut total_vp = Uint128::zero();
+
+    if let Some(voting_escrow) = &cfg.voting_escrow {
+        user_vp = get_voting_power(deps.querier, voting_escrow, account)?;
+        total_vp = get_total_voting_power(deps.querier, voting_escrow)?;
+    }
+
+    let user_virtual_share = user_info.amount.multiply_ratio(4u128, 10u128);
+    let total_virtual_share = pool.total_virtual_supply.multiply_ratio(6u128, 10u128);
```
Comment (epanchee) [inline contracts/tokenomics/generator/src/state.rs:265]: Formula works with real LP tokens amount. Thus we have to use pool's total supply here.

File: packages/astroport/src/generator.rs [hunk 28]
```diff
@@ -162,6 +173,8 @@ pub enum QueryMsg {
     PoolLength {},
     /// Deposit returns the LP token amount deposited in a specific generator
     Deposit { lp_token: String, user: String },
+    /// Returns the current emission boost amount in a specific generator
+    EmissionsBoostRewards { lp_token: String, user: String },
```
Comment (epanchee) [inline packages/astroport/src/generator.rs:177]: Rename to UserVirtualAmount

File: packages/astroport/src/generator.rs [hunk 29]
```diff
@@ -183,6 +196,8 @@ pub enum QueryMsg {
     },
     /// Returns the blocked list of tokens
     BlockedTokensList {},
+    /// Returns the total virtual supply by generator
```
Comment (epanchee) [inline packages/astroport/src/generator.rs:199]: of generator

**Review by epanchee — CHANGES_REQUESTED** [review]
